File: BoxBoxNow/BoxBoxNow/Utilities/Formatters.swift
import SwiftUI

enum Formatters {
    /// Convert ms to M:SS.mmm — matches web msToLapTime()
    static func msToLapTime(_ ms: Double) -> String {
        guard ms > 0 else { return "-" }
        let totalMs = Int(ms)
        let minutes = totalMs / 60000
        let seconds = (totalMs % 60000) / 1000
        let millis = totalMs % 1000
        if minutes > 0 {
            return String(format: "%d:%02d.%03d", minutes, seconds, millis)
        }
        return String(format: "%d.%03d", seconds, millis)
    }

    /// Delta string: +0.123 or -0.456
    static func deltaString(_ ms: Double) -> String {
        let sign = ms >= 0 ? "+" : ""
        return String(format: "%@%.1fs", sign, ms / 1000.0)
    }

    /// Delta color matching web logic
    static func deltaColor(_ ms: Double) -> Color {
        if ms < -10 { return .green }
        if ms > 10 { return .red }
        return .white
    }

    static func speedString(_ kmh: Double) -> String {
        String(format: "%.0f", kmh)
    }

    static func gForceString(_ g: Double) -> String {
        String(format: "%.2f", g)
    }

    /// Race clock HH:MM:SS — matches web format
    static func msToRaceTime(_ ms: Double) -> String {
        guard ms > 0 else { return "--:--:--" }
        let totalSeconds = Int(ms / 1000)
        let hours = totalSeconds / 3600
        let minutes = (totalSeconds % 3600) / 60
        let seconds = totalSeconds % 60
        return String(format: "%d:%02d:%02d", hours, minutes, seconds)
    }

    /// Seconds to HH:MM:SS — matches web secondsToHMS()
    static func secondsToHMS(_ seconds: Int) -> String {
        guard seconds > 0 else { return "00:00:00" }
        let h = seconds / 3600
        let m = (seconds % 3600) / 60
        let s = seconds % 60
        return String(format: "%02d:%02d:%02d", h, m, s)
    }

    /// Seconds to M:SS (short format for stint times)
    static func secondsToStint(_ seconds: Double) -> String {
        guard seconds > 0 else { return "0:00" }
        let min = Int(seconds) / 60
        let sec = Int(seconds) % 60
        return String(format: "%d:%02d", min, sec)
    }

    /// Tier hex color for box score (0-100) — matches web tierHex()
    static func tierColor(_ score: Int) -> Color {
        if score >= 100 { return Color(red: 0.624, green: 0.898, blue: 0.337) } // #9fe556
        if score >= 75  { return Color(red: 0.784, green: 0.914, blue: 0.275) } // #c8e946
        if score >= 50  { return Color(red: 0.898, green: 0.831, blue: 0.227) } // #e5d43a
        if score >= 25  { return Color(red: 0.898, green: 0.604, blue: 0.180) } // #e59a2e
        return Color(red: 0.898, green: 0.267, blue: 0.267) // #e54444
    }

    static func distanceString(_ meters: Double) -> String {
        if meters >= 1000 {
            return String(format: "%.1f km", meters / 1000.0)
        }
        return String(format: "%.0f m", meters)
    }
}

File: android/app/src/main/java/com/boxboxnow/app/util/Formatters.kt
package com.boxboxnow.app.util

import androidx.compose.ui.graphics.Color

object Formatters {
    /** Convert ms to M:SS.mmm */
    fun msToLapTime(ms: Double): String {
        if (ms <= 0) return "-"
        val totalMs = ms.toInt()
        val minutes = totalMs / 60000
        val seconds = (totalMs % 60000) / 1000
        val millis = totalMs % 1000
        return if (minutes > 0) {
            "%d:%02d.%03d".format(minutes, seconds, millis)
        } else {
            "%d.%03d".format(seconds, millis)
        }
    }

    fun deltaString(ms: Double): String {
        val sign = if (ms >= 0) "+" else ""
        return "%s%.1fs".format(sign, ms / 1000.0)
    }

    fun deltaColor(ms: Double): Color = when {
        ms < -10 -> Color(0xFF00C853)
        ms > 10 -> Color(0xFFFF1744)
        else -> Color.White
    }

    fun speedString(kmh: Double): String = "%.0f".format(kmh)
    fun gForceString(g: Double): String = "%.2f".format(g)

    /** Race clock H:MM:SS */
    fun msToRaceTime(ms: Double): String {
        if (ms <= 0) return "--:--:--"
        val total = (ms / 1000).toInt()
        val h = total / 3600
        val m = (total % 3600) / 60
        val s = total % 60
        return "%d:%02d:%02d".format(h, m, s)
    }

    fun secondsToHMS(seconds: Int): String {
        if (seconds <= 0) return "00:00:00"
        val h = seconds / 3600
        val m = (seconds % 3600) / 60
        val s = seconds % 60
        return "%02d:%02d:%02d".format(h, m, s)
    }

    fun secondsToStint(seconds: Double): String {
        if (seconds <= 0) return "0:00"
        val min = seconds.toInt() / 60
        val sec = seconds.toInt() % 60
        return "%d:%02d".format(min, sec)
    }

    /** Tier hex color for box score (0-100) — matches web tierHex() */
    fun tierColor(score: Int): Color = when {
        score >= 100 -> Color(0xFF9FE556)
        score >= 75 -> Color(0xFFC8E946)
        score >= 50 -> Color(0xFFE5D43A)
        score >= 25 -> Color(0xFFE59A2E)
        else -> Color(0xFFE54444)
    }

    fun distanceString(m: Double): String =
        if (m >= 1000) "%.1f km".format(m / 1000) else "%.0f m".format(m)
}
